src/ServiceControl/Recoverability/Retrying/Infrastructure/ReturnToSender.cs

```diff
@@ -30,23 +37,51 @@ public virtual async Task HandleMessage(MessageContext message, IDispatchMessage
 
             if (outgoingHeaders.TryGetValue("ServiceControl.Retry.Attempt.MessageId", out var attemptMessageId))
             {
-                var result = await bodyStorage.TryFetch(attemptMessageId)
-                    .ConfigureAwait(false);
-                if (result.HasResult)
+                if (outgoingHeaders.Remove("ServiceControl.Retry.HasAttachment"))
                 {
-                    using (result.Stream)
+                    var result = await bodyStorage.TryFetch(attemptMessageId)
+                        .ConfigureAwait(false);
+                    if (result.HasResult)
                     {
-                        body = ReadFully(result.Stream);
+                        using (result.Stream)
+                        {
+                            body = ReadFully(result.Stream);
+                        }
+
+                        if (Log.IsDebugEnabled)
+                        {
+                            Log.DebugFormat("{0}: Body size: {1} bytes retrieved from attachment store", messageId, body.LongLength);
+                        }
                     }
-
-                    if (Log.IsDebugEnabled)
+                    else
                     {
-                        Log.DebugFormat("{0}: Body size: {1} bytes", messageId, body.LongLength);
+                        Log.WarnFormat("{0}: Message Body not found in attachment store for attempt Id {1}", messageId, attemptMessageId);
                     }
                 }
                 else
                 {
-                    Log.WarnFormat("{0}: Message Body not found for attempt Id {1}", messageId, attemptMessageId);
+                    string documentId = FailedMessage.MakeDocumentId(outgoingHeaders["ServiceControl.Retry.UniqueMessageId"]);
+                    var results = await documentStore.AsyncDatabaseCommands.GetAsync(new[] { documentId }, null,
+                        transformer: MessagesBodyTransformer.Name).ConfigureAwait(false);
+                    var loadResult = results.Results.SingleOrDefault();
+
+                    if (loadResult != null)
+                    {
+                        string resultBody = ((loadResult["$values"] as RavenJArray)?.SingleOrDefault() as RavenJObject)?.ToObject<MessagesBodyTransformer.Result>()?.Body;
+                        if (resultBody != null)
+                        {
+                            body = Encoding.UTF8.GetBytes(resultBody);
+                        }
+
+                        if (Log.IsDebugEnabled)
+                        {
+                            Log.DebugFormat("{0}: Body size: {1} bytes retrieved from index", messageId, body.LongLength);
```

Comment:
This might need to be moved into the if-statement above. We might end up with

```
MESSAGE-ID: Body size: 0 bytes retrieved from index
```

when in fact, we did not get a value from the index. In any case, we might want to put some more logging in here to catch all of the cases that might come up because they will be a pain to diagnose in support cases.